src/lib/x509/certstor.cpp

```diff
@@ -213,4 +214,135 @@ Certificate_Store_In_Memory::Certificate_Store_In_Memory(const std::string& dir)
    }
 #endif
 
+/**
+ * Internal class implementation (i.e. Pimpl) to keep the required platform-
+ * dependent members of Flatfile_Certificate_Store contained in this compilation
+ * unit.
+ */
+class Flatfile_Certificate_Store_Impl
+   {
+   public:
+      Flatfile_Certificate_Store_Impl(const std::string& system_store) :
+         m_system_store(system_store)
+         {
+
+         if (m_system_store.empty()) { return; }
+
+         DataSource_Stream file(m_system_store);
+         for(const secure_vector<uint8_t> der : PEM_Code::decode_all(file))
+            {
+            m_certificates.push_back(std::make_shared<X509_Certificate>(der.data(), der.size()));
+            }
+         }
+
+   public:
+      const std::string m_system_store;
+      std::vector<std::shared_ptr<const X509_Certificate>> m_certificates;
+   };
+
+Flatfile_Certificate_Store::Flatfile_Certificate_Store() :
+   m_impl(std::make_shared<Flatfile_Certificate_Store_Impl>(""))
+   {
+   }
+
+#if defined(BOTAN_TARGET_OS_HAS_FILESYSTEM)
+Flatfile_Certificate_Store::Flatfile_Certificate_Store(const std::string& file) :
+   m_impl(std::make_shared<Flatfile_Certificate_Store_Impl>(file))
+   {
+   }
+#endif
+
+std::vector<X509_DN> Flatfile_Certificate_Store::all_subjects() const
+   {
+   std::vector<X509_DN> dns;
+   std::transform(m_impl->m_certificates.cbegin(), m_impl->m_certificates.cend(),
+   std::back_inserter(dns), [](std::shared_ptr<const X509_Certificate> cert) { return cert->subject_dn();});
+
+   return  dns;
+   }
+
+std::shared_ptr<const X509_Certificate>
+Flatfile_Certificate_Store::find_cert(const X509_DN& subject_dn,
+                                   const std::vector<uint8_t>& key_id) const
+   {
+   const auto found_cert = std::find_if(m_impl->m_certificates.cbegin(), m_impl->m_certificates.cend(),
+                                        [&](std::shared_ptr<const X509_Certificate> cert)
+      {
+      if(!key_id.empty() && !cert->subject_key_id().empty())
+         {
+         return key_id == cert->subject_key_id() &&
+                subject_dn == cert->subject_dn();
+         }
+      return subject_dn == cert->subject_dn();
+      });
+
+   if(found_cert != m_impl->m_certificates.cend())
+      {
+      return *found_cert;
+      }
+
+   return nullptr;
```

Comment:
Could we re-use this predicate lambda for `find_cert` and `find_all_certs`? Alternatively I think we could implement `find_cert` by `find_all_certs` like so:
```suggestion
const auto certs = find_all_certs(subject_dn, key_id);
return (!certs.empty()) ? certs.front() : nullptr;
```